### PR TITLE
Adds market manager and new interfaces.

### PR DESCRIPTION
**Pull request by murlokito**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -29,16 +29,45 @@ Solnet.Serum is a package within the same `Solnet.` namespace that implements a
...
+    - if the transaction is subject to a custom error defined by the Serum Program [here](https://github.com/project-serum/serum-dex/blob/master/dex/src/error.rs), it is parsed into the appropriate `SerumProgramError` enum value
+- Factory patterns for both `ISerumClient` and `IMarketManager`
+
 
 ## Requirements
 - net 5.0
 
 ## Dependencies
-- Solnet
+- Solnet v0.4.2
```
Comment (tiago18c) [inline README.md:63]: Update

File: Solnet.Serum/IMarketManager.cs [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        /// Initializes the market manager.
+        /// This is an asynchronous operation.
+        /// </summary>
+        Task InitAsync();
+        
+        /// <summary>
+        /// Reloads the associated token accounts and open orders accounts. Useful when an order creates either accounts.
+        /// This is an asynchronous operation.
+        /// </summary>
+        Task ReloadAsync();
```
Comment (tiago18c) [inline Solnet.Serum/IMarketManager.cs:72]: Missing non-async version

File: Solnet.Serum/IMarketManager.cs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        /// <param name="action">An action used to receive the open orders account and the respective slot.</param>
+        /// <returns>A task which performs the operation.</returns>
+        Task SubscribeOpenOrdersAsync(Action<IList<OpenOrder>, ulong> action);
+
+        /// <summary>
+        /// Unsubscribe to the order book of this market.
+        /// This is an asynchronous operation.
+        /// </summary>
+        /// <returns>A task which performs the operation.</returns>
+        Task UnsubscribeOpenOrdersAsync();
```
Comment (tiago18c) [inline Solnet.Serum/IMarketManager.cs:135]: Unsubscribe methods lack non-async overloads

File: Solnet.Serum/ISerumClient.cs [hunk 4]
```diff
@@ -12,11 +13,21 @@ namespace Solnet.Serum
     /// </summary>
     public interface ISerumClient
     {
+        /// <summary>
+        /// The websocket connection state.
+        /// </summary>
+        WebSocketState State { get; }
```
Comment (tiago18c) [inline Solnet.Serum/ISerumClient.cs:19]: Is there any way to be notified of a disconnection from the underlying socket? The subscription object returned from the Subscribe* methods doesn't seem to have any events or other mechanism

File: Solnet.Serum/SerumClient.cs [hunk 5]
```diff
@@ -239,24 +262,24 @@ public Task DisconnectAsync()
...
-                        {
-                            evtQueueSub = subscription;
-                        }
-                    }
 
                     if (evtQueueSub?.Data != null)
                     {
+                        if (!evtQueueSub.TestData)
+                        {
+                            File.WriteAllText("secondInfo.txt", JsonSerializer.Serialize(value,_jsonSerializerOptions));
```
Comment (tiago18c) [inline Solnet.Serum/SerumClient.cs:273]: Gotta remove

File: Solnet.Serum/MarketManager.cs [hunk 6]
```diff
@@ -0,0 +1,874 @@
...
+        /// Initialize the <see cref="Market"/> manager with the given market <see cref="PublicKey"/>.
+        /// </summary>
+        /// <param name="marketAccount">The <see cref="PublicKey"/> of the <see cref="Market"/>.</param>
+        /// <param name="ownerAccount">The <see cref="PublicKey"/> of the owner account.</param>
+        /// <param name="srmAccount">The <see cref="PublicKey"/> of the serum account to use for fee discount, not used when not provided.</param>
+        /// <param name="signatureMethod">A delegate method used to request a signature for transactions crafted by the <see cref="MarketManager"/> which will submit, cancel orders, or settle funds.</param>
+        /// <param name="url">The cluster to use when not passing in a serum client instance.</param>
+        /// <param name="logger">The logger instance to use.</param>
+        /// <param name="serumClient">The serum client instance to use.</param>
+        public MarketManager(PublicKey marketAccount, PublicKey ownerAccount = null, PublicKey srmAccount = null,
```
Comment (tiago18c) [inline Solnet.Serum/MarketManager.cs:124]: We have a factory, make the constructor internal

File: Solnet.Serum/MarketManager.cs [hunk 7]
```diff
@@ -0,0 +1,874 @@
...
+                .CompileMessage();
+
+            byte[] signatureBytes = _requestSignature(txBytes);
+
+            List<byte[]> signatures = new() { signatureBytes };
+            signatures.AddRange(_signers.Select(signer => signer.Sign(txBytes)));
+            _signers.Clear();
+
+            Transaction tx = Transaction.Populate(
+                Message.Deserialize(txBytes), signatures);
```
Comment (tiago18c) [inline Solnet.Serum/MarketManager.cs:438]: can't we save the pre-signature transaction object and just add in the signature?

**Review by tiago18c — APPROVED** [review]
Looks very good.
Just have that minor question regarding re-use of TransactionBuilder

File: Solnet.Serum/MarketManager.cs (hunk 6 above)
Comment (murlokito) [reply Solnet.Serum/MarketManager.cs:124]: Already done with the tests

**Pull request merged**